In [9]:
# Re-import necessary libraries after reset
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Reload the uploaded dataset for the regression model
uploaded_dataset_path = "data/Final Dataset.csv"
final_dataset = pd.read_csv(uploaded_dataset_path)

# Prepare features and target for the regression model
features_final = final_dataset.drop(columns=['close_x'], errors='ignore')  # Replace 'close_x' if target column differs
target_final = final_dataset['close_x']  # Replace with your target column name

# Ensure numeric features and handle missing values
features_final = features_final.select_dtypes(include=[np.number])
features_final.fillna(features_final.mean(), inplace=True)

# Split the dataset into training and testing sets
X_train_final, X_test_final, y_train_final, y_test_final = train_test_split(features_final, target_final, test_size=0.2, random_state=42)

# Train the linear regression model
lr_model_final = LinearRegression()
lr_model_final.fit(X_train_final, y_train_final)

# Predict and evaluate the model
lr_predictions_final = lr_model_final.predict(X_test_final)
lr_mse_final = mean_squared_error(y_test_final, lr_predictions_final)
lr_r2_final = r2_score(y_test_final, lr_predictions_final)

# Display the evaluation metrics
lr_mse_final, lr_r2_final



(0.47142650211741843, 0.9999854539583459)

#### The linear regression model has been evaluated:

Mean Squared Error (MSE): 0.4714 (indicating minimal prediction error).

R² Score: 0.99999 (indicating an excellent fit to the data).

In [39]:
# Remove 'high' and 'low' price features from the dataset
features_reduced = features_final.drop(columns=['high', 'low'], errors='ignore')

# Split the data into training and testing sets
X_train_reduced, X_test_reduced, y_train_reduced, y_test_reduced = train_test_split(
    features_reduced, target_final, test_size=0.2, random_state=42
)

# Train a new linear regression model with the reduced features
lr_model_reduced = LinearRegression()
lr_model_reduced.fit(X_train_reduced, y_train_reduced)

# Predict and evaluate the new model
lr_predictions_reduced = lr_model_reduced.predict(X_test_reduced)
lr_mse_reduced = mean_squared_error(y_test_reduced, lr_predictions_reduced)
lr_r2_reduced = r2_score(y_test_reduced, lr_predictions_reduced)

# Display the updated evaluation metrics
lr_mse_reduced, lr_r2_reduced


(1.6768184904571781, 0.9999482611361498)

In [43]:
from sklearn.linear_model import Ridge, Lasso

# Initialize Ridge and Lasso regression models
ridge_model = Ridge(alpha=1.0, random_state=42)
lasso_model = Lasso(alpha=0.1, random_state=42)

# Train the models using the reduced feature set
ridge_model.fit(X_train_reduced, y_train_reduced)
lasso_model.fit(X_train_reduced, y_train_reduced)

# Predict and evaluate Ridge regression
ridge_predictions = ridge_model.predict(X_test_reduced)
ridge_mse = mean_squared_error(y_test_reduced, ridge_predictions)
ridge_r2 = r2_score(y_test_reduced, ridge_predictions)

# Predict and evaluate Lasso regression
lasso_predictions = lasso_model.predict(X_test_reduced)
lasso_mse = mean_squared_error(y_test_reduced, lasso_predictions)
lasso_r2 = r2_score(y_test_reduced, lasso_predictions)

# Display the results
ridge_mse, ridge_r2, lasso_mse, lasso_r2


(1.6760693456460916, 0.9999482842512941, 1.533053175221027, 0.999952697068908)

In [47]:
# Clean the dataset by replacing invalid values with NaN and imputing missing values
final_dataset_cleaned = final_dataset.replace('.', np.nan)

# Convert all numeric columns to proper types and fill missing values with the mean
for column in final_dataset_cleaned.columns:
    if final_dataset_cleaned[column].dtype == 'object':
        final_dataset_cleaned[column] = pd.to_numeric(final_dataset_cleaned[column], errors='coerce')
final_dataset_cleaned.fillna(final_dataset_cleaned.mean(), inplace=True)

# Re-prepare features and target with cleaned data
features_cleaned = final_dataset_cleaned.drop(columns=['close_x'], errors='ignore')
target_cleaned = final_dataset_cleaned['close_x']

# Remove 'open' feature to reduce R²
features_cleaned_reduced = features_cleaned.drop(columns=['open'], errors='ignore')

# Split the data into training and testing sets
X_train_cleaned, X_test_cleaned, y_train_cleaned, y_test_cleaned = train_test_split(
    features_cleaned_reduced, target_cleaned, test_size=0.2, random_state=42
)

# Train a new linear regression model with cleaned data
lr_model_cleaned = LinearRegression()
lr_model_cleaned.fit(X_train_cleaned, y_train_cleaned)

# Predict and evaluate the cleaned model
lr_predictions_cleaned = lr_model_cleaned.predict(X_test_cleaned)
lr_mse_cleaned = mean_squared_error(y_test_cleaned, lr_predictions_cleaned)
lr_r2_cleaned = r2_score(y_test_cleaned, lr_predictions_cleaned)

# Display the updated evaluation metrics
lr_mse_cleaned, lr_r2_cleaned


(0.5448892752491411, 0.9999831872369096)

In [22]:
# Prepare the full date range from 2004-01 to 2023-12
date_range = pd.date_range(start='2004-01-01', end='2023-12-01', freq='MS')
full_date_df = pd.DataFrame({'Month-Year': date_range.strftime('%Y-%m')})

# Add 'Month-Year' to the dataset
final_dataset['Month-Year'] = final_dataset.apply(
    lambda row: f"{int(row['year'])}-{int(row['month']):02d}", axis=1
)

# Merge the full date range with the dataset to align all months
aligned_data = full_date_df.merge(final_dataset, on='Month-Year', how='left')

# Use 'close_x' as the actual values and predict values for all months
aligned_data['Actual Corn Price'] = aligned_data['close_x']
aligned_features = aligned_data[features_final.columns].fillna(features_final.mean())
aligned_data['Predicted Corn Price'] = lr_model_final.predict(aligned_features)

# Save the results with actual and predicted values
results_file_path = "corn_price_actual_vs_predicted.csv"
aligned_data[['Month-Year', 'Actual Corn Price', 'Predicted Corn Price']].to_csv(results_file_path, index=False)



In [26]:
# Retrieve the coefficients of each feature from the trained linear regression model
feature_coefficients = pd.DataFrame({
    'Feature': features_final.columns,
    'Coefficient': lr_model_final.coef_
})

# Save the feature coefficients to a file
coefficients_file_path = "feature_coefficients.csv"
feature_coefficients.to_csv(coefficients_file_path, index=False)



#### other two methods to analyze dataset
Random Forest Regressor

Gradient Boosting Regressor

In [32]:
# Load the newly uploaded data

                           
uploaded_dataset_path = "data/Final Dataset.csv"
final_dataset = pd.read_csv(uploaded_dataset_path)

# Prepare features and target from the uploaded data
features_new = final_dataset.drop(columns=['close_x'], errors='ignore')
target_new = final_dataset['close_x']

# Ensure numeric features only and handle missing values
features_new = features_new.select_dtypes(include=[np.number])
features_new.fillna(features_new.mean(), inplace=True)

# Split into training and testing sets
from sklearn.model_selection import train_test_split
X_train_new, X_test_new, y_train_new, y_test_new = train_test_split(
    features_new, target_new, test_size=0.2, random_state=42
)

# Explore alternative models: Random Forest and Gradient Boosting
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Initialize models
rf_model = RandomForestRegressor(random_state=42)
gb_model = GradientBoostingRegressor(random_state=42)

# Train models
rf_model.fit(X_train_new, y_train_new)
gb_model.fit(X_train_new, y_train_new)

# Predict and evaluate models
rf_predictions = rf_model.predict(X_test_new)
gb_predictions = gb_model.predict(X_test_new)

rf_mse = mean_squared_error(y_test_new, rf_predictions)
rf_r2 = r2_score(y_test_new, rf_predictions)

gb_mse = mean_squared_error(y_test_new, gb_predictions)
gb_r2 = r2_score(y_test_new, gb_predictions)

rf_mse, rf_r2, gb_mse, gb_r2


(9.182760035158775, 0.9997166624927312, 4.904293871803718, 0.9998486761719537)